src/pybind/mgr/dashboard/services/nvmeof_conf.py

```diff
@@ -55,13 +55,19 @@ def add_gateway(cls, name, service_url, group, daemon_name):
         config = cls.get_gateways_config()
 
         if name in config.get('gateways', {}):
-            existing_gateways = config['gateways'][name]
-            for gateway in existing_gateways:
-                if 'daemon_name' not in gateway:
-                    gateway['daemon_name'] = daemon_name
-                    break
-                if gateway['service_url'] == service_url:
-                    return
+            # the nvmeof dashboard config used in v19.2.0 saves the below
+            # to a dict. Converting that to a list so that the upgrade
+            # properly migrate it to the newer format, and also keep it empty.
+            if isinstance(config['gateways'][name], dict):
+                config['gateways'][name] = []
```

Comment:
So I understand that in version prior to 19.2.0 the gateways are saved as a dict, and from 19.2.0 onwards a list is used. Is that right?

What I don't get is why is this being emptied here, what would happen if you are upgrading from a newer version? What's the purpose of resetting this value?

Reply:
> So I understand that in version prior to 19.2.0 the gateways are saved as a dict, and from 19.2.0 onwards a list is used. Is that right?

@Pegonzal yup, that's right. One thing, nvmeof was introduced in 19.2.0, and the structure is changed in 19.2.1 as you said.

its emptied here and immediately being set below using the `new_gateway` dict. During upgrade, the nvmeof daemons are redeployed and everytime a redeploy happens this particular code is triggered by cephadm. So I create a list (by removing the pre-existing configuration because its invalid), and then keeps on adding new daemons to this new list as cephadm redeploys each daemons.

Reply:
Okay, thanks for the explanation @nizamial09 .

One last question here:

- The issue of different type formats it's now clear, but why do we want to empty it if it's a dict, shouldn't the data that contained be transformed to a list? Instead of just removing it

Reply:
hmm yeah, you are right. I was only looking at the possibility of cephadm doing this for us but someone with an older configuration can still try to change it manually and their config can be lost. But I am not really concerned about that tbh, because most of the users should already be migrated to the newer configs already as part of the 19.2.1 (although manually). And the case of dict is rarer now and only during upgrade.. And if they have upgraded to a newer version (which includes this fix) from the one that has the bug, again cephadm takes care of the whole thing and we don't need to worry about the onetime dict thing. Which is why I didn't go for the extra code that is only needed at the extreme scenario.